File: last-stone-weight/last-stone-weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones = [-n for n in stones]
        heapq.heapify(stones)
        n = 0
        while len(stones) > 1 or (len(stones) == 1 and n != 0):
            if n == 0:
                n1 = heapq.heappop(stones)
            else:
                n1 = heapq.heappushpop(stones, n)
            n2 = heapq.heappop(stones)
            n = n1-n2
        if stones:
            return -stones[0]
        else:
            return -n

File: last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        make_heap(stones.begin(), stones.end());
        int x, y;
        while(stones.size() > 1){
            y = stones[0]; 
            pop_heap(stones.begin(), stones.end()); stones.pop_back();
            x = stones[0];
            pop_heap(stones.begin(), stones.end()); stones.pop_back();
            if (x == y) continue;
            stones.push_back(y - x);
            push_heap(stones.begin(), stones.end());
        }
        if(stones.empty()) return 0;
        return stones[0];
    }
};